### ISSUE TITLE
RFC: torch.vmap

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#42368]
TL;DR: We’re implementing a prototype torch.vmap (inspired by jax.vmap) for PyTorch and are looking for early adopters! Please go to https://github.com/pytorch/functorch and follow installation instructions there.

JAX provides jax.vmap, a program transformation that automatically “vectorizes” a function. What this means is that it parallelizes a function over some dimension of the inputs. jax.vmap can potentially do the following:

* Automatically batch certain models (#830). vmap can help users forget about the batch dimension, enabling a simpler model authoring experience.
* Efficiently compute batched gradients (#8304, #23475). vmap can speed up things like full Jacobian/Hessian matrix computation.
* Efficiently compute per-sample-gradients (#7786).
* And more (see [jax’s README](https://github.com/google/jax#auto-vectorization-with-vmap))

Given the popularity of JAX, we’re implementing a prototype of vmap in PyTorch to see how well it applies to PyTorch use cases. Because the torch.vmap project is in early stages and requires some investment, we’d love to hear from and work directly with early adopters to shape the initial direction.

[UPDATE 11/12/21]
Are you running PyTorch 1.10 and want to use this? See installation instructions at https://github.com/pytorch/functorch

[UPDATE 5/5/21]
 We have a working prototype of this (JAX-like composable vmap + grad for PyTorch) and it currently lives out-of-tree at https://github.com/pytorch/functorch.

We’re very interested in supporting your use cases to build out and validate the prototype! If you’re not already in contact with us, please open an issue at https://github.com/facebookresearch/functorch if you’re interested in using the prototype. You can also try it out via pip install --user "git+https://github.com/pytorch/functorch.git"

Alternatively, please:
- post in the [channel for functorch users](https://app.slack.com/client/T2077MDKQ/C02DVE5MAFR) if you're on the PyTorch slack
- or reach out directly to me, on the PyTorch Slack channel or via email (my github username @gmail.com) if you’d be interested in using torch.vmap for your project.

**Comment by manuelbaum** [comment]
Thanks, I think this is a really, really useful feature!

Is it possible that it doesn't work with torch.autograd.jacobian yet?

If I try this, I get an error:

`nbatch = 4
A = torch.randn(nbatch, 2,2)
f = lambda x: x**2

J = torch.autograd.functional.jacobian(f, (A[0]))
lambda_jac = lambda x: torch.autograd.functional.jacobian(f, (x))
vmapped_jacobian = torch.vmap(lambda_jac)

J_vmapped = vmapped_jacobian(A)
print(J_vmapped)`

and this is the error I get:

`Traceback (most recent call last):
  File "/home/manuelbaum/ws_domip/src/domip/domip/playground/batch_jacobian.py", line 33, in <module>
    J_vmapped = vmapped_jacobian(A)
  File "/home/manuelbaum/.local/lib/python3.8/site-packages/torch/_vmap_internals.py", line 262, in wrapped
    batched_outputs = func(*batched_inputs)
  File "/home/manuelbaum/ws_domip/src/domip/domip/playground/batch_jacobian.py", line 30, in <lambda>
    lambda_jac = lambda x: torch.autograd.functional.jacobian(f, (x))
  File "/home/manuelbaum/.local/lib/python3.8/site-packages/torch/autograd/functional.py", line 423, in jacobian
    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
  File "/home/manuelbaum/.local/lib/python3.8/site-packages/torch/autograd/functional.py", line 61, in _grad_preprocess
    res.append(inp.detach().requires_grad_(need_graph))
RuntimeError: Batching rule not implemented for aten::detach; the fallback path doesn't work on out= or view ops.`

I am using torch version 1.8.0.dev20201018

Please tell me if I am doing something wrong. Otherwise, it would be really great if you implemented it also to work for jacobians.

**Comment by zou3519** [comment]
> Thanks, I think this is a really, really useful feature!
> 
> Is it possible that it doesn't work with torch.autograd.jacobian yet?

torch.vmap doesn't work with torch.autograd.jacobian yet. I'll add it to the list of things to investigate. Could I ask about what your use case for batched jacobians is?

**Comment by manuelbaum** [comment]
they would come in pretty handy because i am implementing a differentiable extended kalman filter. having batch jacobian computation in there would help with 

1.) batch training the models in that filter, because here jacobian computation is actually part of the algorithm that i want to learn parameters for.
2.) e.g. for 6D EKF slam on a [D = 6 + 3*n_landmarks] dimensional state, the full state-transition jacobian is DxD dimensional and takes rather long to compute with many landmarks, but I also know that the jacobian is pretty sparse, blockdiagonal. So I could compose that larger jacobian out of smaller jacobian blocks because I know the structure of that matrix.

Both of these cases could be solved by looping of course, but especially because the model is supposed to work in realtime with many landmarks a batch solution would be really helpful.

**Comment by jackeown** [comment]
One nice thing to have here would be a vmap wrapper for modules as well as functions.
I'm not sure if this fully makes sense, but I'm thinking something where you could say `myModule = torch.vmap(myModule)` and it will become a module where the forward method is vmapped.

Could this be possible?

**Comment by zou3519** [comment]
> One nice thing to have here would be a vmap wrapper for modules as well as functions.
> I'm not sure if this fully makes sense, but I'm thinking something where you could say `myModule = torch.vmap(myModule)` and it will become a module where the forward method is vmapped.
> 
> Could this be possible?

Thanks for the suggestion, @jackeown. In the end state we would like this to work on modules, but the details and API aren't completely clear at the moment.

**Comment by tellihichem** [comment]
Hi all,
Is it possible of adding batching rule for **aten::einsum** in the near future !?
My first try for _torch.vmap_ on special function I wrote was this error. I'm glade that there is some developers whom sacrifices there time for adding cool function to this awsome library. this batch vectorization _torch.vmap_ is one of the biggest features that will enrich _pytoch_. I hope it will be generilized across all _aten_ functions.
Good luck.
`RuntimeError: Batching rule not implemented for aten::einsum. We could not generate a fallback.`

**Comment by sirmarcel** [comment]
Hello! Just wanted to drop in to say that `vmap` has some quite interesting applications when it comes to computing jacobian-related properties in the context of NNs for physics, such as [SchNet](https://github.com/atomistic-machine-learning/schnetpack), so I'm really quite keenly following the developments here. Is there any chance this feature will make it into the next release of `pytorch`?

**Comment by zou3519** [comment]
@sirmarcel does a faster [torch.autograd.functional.jacobian](https://pytorch.org/docs/stable/autograd.html#torch.autograd.functional.jacobian) serve your use case or is there something specific that you'd want to do with a vmap? We're planning on adding a flag to torch.autograd.functional.jacobian to leverage torch.vmap in the backend (instead of invoking the autograd engine multiple times as it currently does).

**Comment by AlphaBetaGamma96** [comment]
Hello! Following on from what @sirmarcel said, are there any plans to extend this for [torch.autograd.functional.hessian](https://pytorch.org/docs/stable/autograd.html#torch.autograd.functional.hessian) as well? I, too, am using neural network for physics application and calculating the hessian is required when calculating certain properties. My current method for calculating it, is enumerating over my input and calculating the hessian one sample at a time and storing them within a new Tensor. The idea of vectorizing such an operation would be a significant speed-up. Is there a chance that `vmap` will be added to `torch.autograd.functional.hessian` too?

**Comment by zou3519** [comment]
@AlphaBetaGamma96, yes, we have plans to extend this for both torch.autograd.functional.hessian and torch.autograd.functional.jacobian.  I am not sure the feature will be in the 1.8 release, though -- one of the blockers is demonstrating stability and good performance on existing use cases.

 It's possible (and very likely!) that in torch.vmap's current state, it already provides speedups toward your use case. If you're feeling adventurous (and would like to provide feedback), here's [an example for how to use it to compute efficient jacobians or hessians](https://gist.github.com/zou3519/b8c0c534c7509e7a3ed0612d01226fb7). Currently, the torch.vmap API is only in PyTorch nightly binaries ([download instructions here](https://pytorch.org)).

Please feel free to reply here or ping me on [Slack](https://pytorch.slack.com/) if you want to chat more!

**Comment by AlphaBetaGamma96** [comment]
@zou3519 I'm currently working with 1.6, so I'll update PyTorch to 1.7 and have a look at `torch.vmap` and see if I can get it working and notice any improvement in my code! (Or should I go straight to 1.8, or is there no difference between `torch._vmap_internals` and `vmap` in 1.8?) Thank you!

**Comment by zou3519** [comment]
> @zou3519 I'm currently working with 1.6, so I'll update PyTorch to 1.7 and have a look at `torch.vmap` and see if I can get it working and notice any improvement in my code! (Or should I go straight to 1.8, or is there no difference between `torch._vmap_internals` and `vmap` in 1.8?) Thank you!

1.8 isn't out yet, so trying to use `from torch._vmap_internals import vmap` in PyTorch 1.7 would be great for experimentation. `torch.vmap` in the PyTorch nightly releases has more performance improvements.

`vmap` should raise a warning whenever a batching rule (necessary for performance) isn't implemented -- if you let us know which operators `vmap` warns on for your use case, we can work on prioritizing speeding those up :)

**Comment by sirmarcel** [comment]
> @sirmarcel does a faster [torch.autograd.functional.jacobian](https://pytorch.org/docs/stable/autograd.html#torch.autograd.functional.jacobian) serve your use case or is there something specific that you'd want to do with a vmap? We're planning on adding a flag to torch.autograd.functional.jacobian to leverage torch.vmap in the backend (instead of invoking the autograd engine multiple times as it currently does).

The quantity I need to compute  is derived from the Jacobian (I need to do products/sums across some dimensions with other tensors). So in principle the Jacobian is enough, since I can do all the other steps later. It might be slightly inefficient because I need to store the entire Jacobian in memory before reducing it down, but that's probably ok. 

Currently I'm looping over the output dimension in `python`, which is really very slow. If `vmap` works as expected and speeds it up significantly, this would indeed be a very good usecase :)

If it's already experimentally present in 1.7, I might even be able to try this out ...

**Comment by zou3519** [comment]
@sirmarcel thank you for the context!

> Currently I'm looping over the output dimension in `python`, which is really very slow. If `vmap` works as expected and speeds it up significantly, this would indeed be a very good usecase :)
> 
> If it's already experimentally present in 1.7, I might even be able to try this out ...

Yes, it is experimentally present in 1.7: see https://gist.github.com/zou3519/b8c0c534c7509e7a3ed0612d01226fb7.

**Comment by sirmarcel** [comment]
@zou3519 great! i'll have chat with our HPC people if we can set that up. they're reluctant to instal nightlies for me ;)

**Comment by sirmarcel** [comment]
@zou3519 I've tried it locally (CPU only, for implementing), it seems like `aten::_index_put_impl_` is missing a batching rule. Is there an easy way  to tell which python api operation this corresponds to in the forward pass? I might be able to hack this out of the NN, but I don't know where to start... (I'm assuming this won't be implemented soon, judging from [this issue](https://github.com/pytorch/pytorch/issues/49562).)

**Comment by zou3519** [comment]
@sirmarcel, let's open up another issue (we can title it "`aten::_index_put_impl_` missing batching rule" or something) to discuss your use case further? I'm curious to learn about what other batching rules are missing for your use case. Regarding your question: `_index_put_impl_` gets invoked during advanced indexing (e.g., `tensor[index] = other_tensor`) or if [`index_put_` is called](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.index_put_).

**Comment by AlphaBetaGamma96** [comment]
Hi @zou3519, with the [example](https://gist.github.com/zou3519/b8c0c534c7509e7a3ed0612d01226fb7) you posted above for using batching with the jacobian and hessian calculations.

Firstly, does `vmap` work directly with `torch.autograd.functional.hessian`? Rather than calling `torch.autograd.grad` twice, or is this to be implemented in the future as was mentioned in your previous reply? Because I've tried my code with using `torch.autograd.functional.hessian` and calling `torch.autograd.grad` twice and there's about an order of magnitude difference in their runtimes. 

Secondly, when calling this batched form of the `jacobian` or the `hessian` does this batch get remembered within the backward pass? The reason I'm asking about this is that I had an error within my code that was due to a mismatch in dimensions which emerged from one `tensor` (a) containing the batch dimension and another `tensor` (b) being of the same size but each element within the `tensor` contained a batch of `1` because those elements were calculated via `torch.autograd.functional.hessian` which is restricted to a batch of `1` within pytorch 1.6. To give a clearer idea of what I'm talking about I recently posted a question about this on the forums [here](https://discuss.pytorch.org/t/how-to-visualize-backward-and-perhaps-doublebackward-pass-of-variable/108456/16). This realisation happens at reply #16

Thank you for the helpful discussion! :)

**Comment by zou3519** [comment]
@AlphaBetaGamma96 let's hop on a thread on the PyTorch forums to discuss further (I think we'll have a bit to discuss and I don't want to have extended discussions directly on this issue). I'll send you a message on the forums

**Comment by Zymrael** [comment]
Hi @zou3519, thanks for the great work done so far. Beyond standard use cases, I've been attempting `vmapped` forwards where each elements in a sequence of inputs `X` gets passed to a specific module in a list or `nn.Sequential`. The pseudocode for this would be:
```
def vmap_forward(X, layers):
    def get_forward(z0):
        return layers[idx](x)
    return vmap(get_forward)(x)
```
where the issue becomes how to specify `idx`.  Naively defining and vmapping over an additional input of indexes raises `Batching rule not implemented for aten::item`
```
def vmap_forward(X, layers):
    def get_forward(z0, idx):
        return layers[idx](x)
    return vmap(get_forward)(Z, idxs)
```
I was wondering if there is a recommended approach for this that I haven't been able to figure out, any help would be appreciated.

**Comment by manuelbaum** [comment]
> @AlphaBetaGamma96, yes, we have plans to extend this for both torch.autograd.functional.hessian and torch.autograd.functional.jacobian. I am not sure the feature will be in the 1.8 release, though -- one of the blockers is demonstrating stability and good performance on existing use cases.
> 
> It's possible (and very likely!) that in torch.vmap's current state, it already provides speedups toward your use case. If you're feeling adventurous (and would like to provide feedback), here's [an example for how to use it to compute efficient jacobians or hessians](https://gist.github.com/zou3519/b8c0c534c7509e7a3ed0612d01226fb7). Currently, the torch.vmap API is only in PyTorch nightly binaries ([download instructions here](https://pytorch.org)), but you can also try to `from torch._vmap_internals import vmap` if you have PyTorch 1.7.
> 
> Please feel free to reply here or ping me on [Slack](https://pytorch.slack.com/) if you want to chat more!

@zou3519 as far as i can see, the jacobian computation in that linked example is "just" a sped up version that uses vmap internally (as with the vectorize to torch.autograd.functional.jacobian flag that came with 1.8). Would it be possible to also compute *batch* jacobian computation that way? it seems like many people could use this and this could maybe be a quick way to get it done?

**Comment by zou3519** [comment]
> Would it be possible to also compute batch jacobian computation that way? it seems like many people could use this and this could maybe be a quick way to get it done?

@manuelbaum we're looking into it! Right now there isn't a way to use the torch.vmap prototype to compute batch jacobian, but that's on the list of functionality we want to provide.

**Comment by zou3519** [comment]
@manuelbaum we're getting close to having a prototype where torch.vmap can be used to compute batch jacobian. To validate the prototype, we've been testing it out on different examples. Do you have an example of a function that you'd like to compute batch-jacobian for? (Also, are you on slack and/or the PyTorch forums? Trying not to have too many single-threaded conversations on this issue)

**Comment by muhrin** [comment]
Hi @zou3519 , sorry to spam this thread further but I haven't had my slack request approved yet.  Just adding to the chorus of people asking for batch `jacobian` support.  In general for physics applications you often want to train a model to fit both a function _and_ its derivative (and sometimes second derivative).  This is where the Jacobian (and Hessian) calculation comes in: you use it to calculate the derivatives of the function represented by the model and then use this as part of the loss function to make the model learn both the value and derivatives simultaneously.  The canonical example is a energy function which you also want to fit to the forces (the, negative, first derivative of the energy wrt positions).

Without batch Jacobian the only way I've found to do this is 1) loop over all inputs calling `jacobian` which is slow, or, 2) call `jacobian` using an entire batch of inputs knowing that this will produce a block diagonal matrix that is highly spare.  Unfortunately for me, while fast, 2) blows through my memory so I'm stuck with a painfully slow 1).

Happy to give a code example or discuss more if that's useful.

**Comment by smorad** [comment]
Is this stackable? I'm working on recurrent models, so it would be very nice to `vmap` across both batch and time dimensions: `vmap(vmap(my_model))`. I'm guessing this does not work with ragged tensors, and so everything must be padded to the correct size?

E.g. of ragged tensor usage:
```
A = torch.arange(3*4*5).reshape(3,4,5)
batches = A.shape[0]
timesteps = A.shape[1]
num_timesteps = torch.arange(timesteps)
batch_idxs = torch.arange(batches)

# Compute results for latest timestep
A[batch_idxs, num_timesteps[batch_idxs]] = ...
```

**Comment by zou3519** [comment]
@smorad this is stackable but yes, it does not work with raggedtensors, so everything must be padded to the correct size

**Comment by sprajagopal** [comment]
@zou3519 I read through the docs [here](https://pytorch.org/docs/master/generated/torch.vmap.html). Trying to understand how `vmap` works. Tried this:
``` py
import torch
def generate_scalars(c):
     scalar = 1 if random() > 0.5 else -1
     return torch.tensor(scalar) 

A = torch.randn(2,3)

for i in range(10):
     print(vmap(generate_scalars)(A))
```
I assumed `vmap` would apply `generate_scalar` on each row of A. So `random()` would probably give `1` and `-1` at some point.
But the output was this:
```
tensor([-1, -1])
tensor([1, 1])
tensor([-1, -1])
tensor([1, 1])
tensor([1, 1])
tensor([1, 1])
tensor([-1, -1])
tensor([1, 1])
tensor([-1, -1])
tensor([1, 1])
```
I tried the function a 100 times as well, but the pattern is consistent. I think I misunderstood some part of this.

**Comment by awf** [comment]
Should we expect this to reach TorchScript?   I.e. today annotating a function using vmap with `@script` produces an error:
```python
import torch

def cthulu(x : float) -> float:
  if x < 0.0:
    return 0.0
  elif x < 1.0:
    return 1/3 * x ** 3
  else:
    return x - 2/3

@torch.jit.script
def vcthulu(x : torch.Tensor):
  return torch.vmap(cthulu, x)
```

```
# python examples/dl-activations/cthulu.py 
...
RuntimeError: 
Unknown type name 'Callable':
  File "/anaconda/envs/ptnightly/lib/python3.8/site-packages/torch/_vmap_internals.py", line 145
def vmap(func: Callable, in_dims: in_dims_t = 0, out_dims: out_dims_t = 0) -> Callable:
               ~~~~~~~~ <--- HERE
    """
    vmap is the vectorizing map. Returns a new function that maps `func` over some
'vmap' is being compiled since it was called from 'vcthulu'
  File "examples/dl-activations/cthulu.py", line 14
@torch.jit.script
def vcthulu(x : torch.Tensor):
  return torch.vmap(cthulu, x)
         ~~~~~~~~~~~~~~~~~~~~ <--- HERE
```

**Comment by zou3519** [comment]
> Should we expect this to reach TorchScript? I.e. today annotating a function using vmap with @script produces an error:

This is not expected to work with TorchScript right now

**Comment by zou3519** [comment]
A quick update for those of you who are following this thread:

We're actively developing a prototype of this (JAX-like composable vmap + grad for PyTorch) that current works on many interesting examples. The prototype lives out-of-tree at https://github.com/zou3519/functorch.

We’re very interested in supporting your use cases to build out and validate the prototype! If you’re not already in contact with us, please open an issue at https://github.com/zou3519/functorch if you’re interested in using the prototype.

**Comment by tomsons22** [comment]
Hi,

Quick question for vmap. I'm using the nightly version of pytorch (installed today). I want to vmap a function that calls autograd.grad inside. Here's a small example:

```
def f(x):
    return torch.dot(x, x)
def gf(x):
    return torch.autograd.grad(f(x), x)[0]
def grads_vmap(X):
    return vmap(gf)(X)
X = torch.rand(100, 10, requires_grad=True)
gs = grads_vmap(X)
```

This does not work, I get the error 

> RuntimeError: torch.autograd.grad(outputs, inputs, grad_outputs) called inside torch.vmap. We do not support the case where any outputs are vmapped tensors (output 0 is being vmapped over). Please call autograd.grad() outside torch.vmap or file a bug report with your use case.

I guess vmap in its current version does not work with functions that call grad inside it? Or am I doing something wrong? I know that in this case there are several workarounds, this is just a small example that shows the issue.

Thanks!

**Comment by zou3519** [comment]
@tomsons22 we've moved development of vmap over to https://github.com/zou3519/functorch. As the error suggests, we do not support calling autograd.grad inside of vmap.

However, functorch has separate `vmap` and `grad` transforms and you can indeed call the `grad` transform inside of `vmap`. Your example would look like the following:
```
import torch
from functorch import grad, vmap

def f(x):
    return torch.dot(x, x)

gf = grad(f)

def grads_vmap(X):
    return vmap(gf)(X)
X = torch.rand(100, 10, requires_grad=True)
gs = grads_vmap(X)
```

**Comment by catskillsresearch** [comment]
Here is a good use case, vectorizing linear regression on a sequence of X-Y pair datasets of different lengths.  Goal is to ship every element of batch to GPU to run in full parallel: https://github.com/tensorflow/tensorflow/issues/51771

**Comment by DiffeoInvariant** [comment]
We have a use for this in (CRIKit)[https://gitlab.com/crikit/crikit] -- we're a library for inference in materials science, and one of the very nice features of JAX is that with `vmap`, we can write material models in a "pointwise" manner, i.e. write the model as if it takes in inputs one point at a time, but evaluate it on a whole mesh all at once. We currently support models written in JAX and TensorFlow, and would like to add PyTorch support, but it would be really awkward and likely inefficient without `torch.vmap`. In particular, it would be great if `torch.vmap` presented a similar API to `jax.vmap`, with the `in_axes` and `out_axes` parameters.

**Comment by zou3519** [comment]
> We have a use for this in (CRIKit)[https://gitlab.com/crikit/crikit] -- we're a library for inference in materials science, and one of the very nice features of JAX is that with `vmap`, we can write material models in a "pointwise" manner, i.e. write the model as if it takes in inputs one point at a time, but evaluate it on a whole mesh all at once. We currently support models written in JAX and TensorFlow, and would like to add PyTorch support, but it would be really awkward and likely inefficient without `torch.vmap`. In particular, it would be great if `torch.vmap` presented a similar API to `jax.vmap`, with the `in_axes` and `out_axes` parameters.

@DiffeoInvariant we have a new and improved vmap that works with PyTorch over at https://github.com/facebookresearch/functorch. In a way that is analogous to `jax.vmap`, `funtorch.vmap` has `in_dims` and `out_dims` parameters. Please give it a try, and if you run into issues please do open us an issue to chat more on our [github issues tracker](https://github.com/facebookresearch/functorch/issues)

**Comment by PeaBrane** [comment]
I'm not sure if you are interested in gathering use cases.

But at least for my work in object detection, having the vmap functionality would be a godsend. This is especially true for batched and vectorized bounding box assignment to anchor boxes in the data-preprocessing stage, and bounding box suppression (non-max) in the post-processing stage.

Take non-max suppression for example, where the objective is to compute the IOUs (overlaps) between each pair of bounding boxes, and then throw away the ones with high overlaps to the high confidence bounding box. The code usually gets very messy especially if you have multiple classes, multiple frames, and multiple anchor-box aspect ratios. 

It would be very nice to have a function to just deal with non-max suppression on a single frame and single class, and have it immediately generalizable to the more complicated cases with a vmap higher-order function

**Comment by edgarriba** [comment]
@PeaBrane exactly what we are waiting for here :)
https://github.com/kornia/kornia/blob/master/kornia/geometry/bbox.py#L495

**Comment by zou3519** [comment]
@PeaBrane @edgarriba Thanks for the interest. I don't completely follow, but, is there a code snippet with example inputs that shows what you do today?

**Comment by PeaBrane** [comment]
@zou3519 Hi, unfortunately I cannot provide my code because it's for a company, and usually the code for non-max suppression (NMS) is not terribly illuminating either.

But the concept of NMS is simple and quite standard in CV, if you imagine a single image, and a bunch of predicted bounding boxes in it, sometimes the outputted bounding boxes overlap (say the network outputs two separate bounding boxes for a single car). This is not desirable and you need a way to remove the redundant bounding boxes. The procedure for doing is to take the highest confidence bounding box, and then remove the other ones with a high intersection with the high-confidence box, and you iterate this algorithm until there is no boxes to remove (or when there is no more overlapping boxes).

The implementation of this algorithm is not terribly difficult for a single image with a single class, but the code gets very messy if you want to vectorize NMS for simultaneous box removals on multiple frames and for multiple classes. I think having vmap would help in the sense that, the coder can just focus on implementing NMS properly for a single image, and then vmap is going to take care of extending this for simultaneous NMS for multiple images.

NMS is just an example (and I think it's actually a poorly chosen one...) of a CV processing task, and I think vmap can help most image processing tasks in general.

**Comment by macio232** [comment]
I am trying to extend `ImplicitFnGradientRerouting` (https://github.com/rtqichen/torchdiffeq/blob/8df757cb12f231a6b4349a96608b7a9d11166988/torchdiffeq/_impl/odeint.py#L130) for batching.

The problem I get is that gradients returned by `vjp` (form pytorch, not functorch) are incorrect - always 0. When I execute `grad` ( from pytorch) on `BatchedTensor`s that appear with `vmap`, I get:
`*** RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn`
what is not true.

Is there something wrong on my side, or is this the expected behavior/not yet supported use case?

**Comment by macio232** [comment]
I experience `torch.autograd.functional.vjp` giving incorrect gradients (always 0) when called on a `BatchedTensor` object in a `vmap`ed function.

Is this a bug or feature?

**Comment by zou3519** [comment]
@macio232 do you have a repro? Please feel free to open an new issue over at https://github.com/pytorch/functorch so we can discuss in another thread

**Comment by JustinS6626** [comment]
Hi @zou3519, I am trying to use the `vmap` function so that it not only batches the main input, but it also batches a sequence of integer values, with each instance in the ensemble using value that it gets to subscript a list that it receives in another argument. Here is a toy example of what I am attempting:

```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import vmap
from torch.func import stack_module_state
from torch.func import functional_call
import copy
import numpy as np
torch.manual_seed(0)


class SimpleMLP(nn.Module):
    def __init__(self):
        super(SimpleMLP, self).__init__()
        self.fc1 = nn.Linear(784, 128)
        self.fc2 = nn.Linear(128, 128)
        self.fc3 = nn.Linear(128, 10)

    def forward(self, x, letter_index, letters):
        letter_choice = letter_index.clone().cpu().numpy()
        print(letters[letter_choice])
        x = x.flatten(1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.fc2(x)
        x = F.relu(x)
        x = self.fc3(x)
        return x


device = 'cuda'
num_models = 10
letter_list = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
data = torch.randn(100, 64, 1, 28, 28, device=device)
targets = torch.randint(10, (6400,), device=device)
minibatches = data[:num_models]

models = [SimpleMLP().to(device) for _ in range(num_models)]
letter_nums = np.arange(10)
letter_index = torch.tensor(letter_nums).to(device)

params, buffers = stack_module_state(models)

base_model = copy.deepcopy(models[0])
base_model = base_model.to(device)

def fmodel(params, buffers, x, index, markers):
    return functional_call(base_model, (params, buffers), (x, index, markers))

predictions1_vmap = vmap(fmodel, in_dims=(0, 0, 0, 0, None))(params, buffers, minibatches, letter_index, letter_list)

```

When I run it, I get the following error:

```
Traceback (most recent call last):
  File "/usr/lib/python3.11/idlelib/run.py", line 578, in runcode
    exec(code, self.locals)
  File "/home/owner/ensemble_toy_example.py", line 57, in <module>
    predictions1_vmap = vmap(fmodel, in_dims=(0, 0, 0, 0, None))(params, buffers, minibatches, letter_index, letter_list)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/_functorch/apis.py", line 188, in wrapped
    return vmap_impl(func, in_dims, out_dims, randomness, chunk_size, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/_functorch/vmap.py", line 278, in vmap_impl
    return _flat_vmap(
           ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/_functorch/vmap.py", line 44, in fn
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/_functorch/vmap.py", line 391, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/owner/ensemble_toy_example.py", line 55, in fmodel
    return functional_call(base_model, (params, buffers), (x, index, markers))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/_functorch/functional_call.py", line 143, in functional_call
    return nn.utils.stateless._functional_call(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/utils/stateless.py", line 263, in _functional_call
    return module(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/owner/ensemble_toy_example.py", line 20, in forward
    letter_choice = letter_index.clone().cpu().numpy()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
```

I am therefore wondering if there is something that I can do to get this approach to work properly in `vmap`. Alternatively, would it be possible, in the next update, to enable calls to `Tensor.item()' within `vmap`? I think that would greatly extend the flexibility of `vmap` to allow for a broader range of ensemble operations.

Thank you very much!

**Comment by KinWaiCheuk** [comment]
@zou3519 
I need to calculate [per-sample-gradient](https://pytorch.org/tutorials/intermediate/per_sample_grads.html) using `vmap` in my projects very often.

However, `vmap` doesn't work when the model uses `torch.utils.checkpoint`.

Here's a dummy code that represents my use case
```python
import torch
import torch.nn as nn
import torch.optim as optim
import random
from torch.func import vmap
from torch.utils.checkpoint import checkpoint

# Set a fixed seed for reproducibility
seed = 42
torch.manual_seed(seed)
random.seed(seed)
if torch.cuda.is_available():
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)  # For multi-GPU setups
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

# Define a simple Transformer model
class SimpleTransformer(nn.Module):
    def __init__(self, input_dim, model_dim, num_heads, num_layers, output_dim):
        super(SimpleTransformer, self).__init__()
        self.embedding = nn.Embedding(input_dim, model_dim)
        self.transformer = nn.Transformer(d_model=model_dim, nhead=num_heads, num_encoder_layers=num_layers)
        self.fc_out = nn.Linear(model_dim, output_dim)
        
    def forward(self, src, tgt):
        src = self.embedding(src)
        tgt = self.embedding(tgt)
        # output = self.transformer(src, tgt)
        output = checkpoint(self.transformer, src, tgt, use_reentrant=False)
        return self.fc_out(output)
    
def compute_loss(model,
                 weights,
                 buffers,
                 src,
                 tgt,
                 ):

    # Forward pass
    output = torch.func.functional_call(
        model, (weights, buffers), args=(src.unsqueeze(1), tgt[:-1].unsqueeze(1))
        )

    # Reshape output and target for loss calculation
    output = output.view(-1, output_dim)
    tgt = tgt[1:].view(-1)  # Shift target sequence by one
    loss = criterion_mean(output, tgt)
    print(loss.shape)
    return loss

# Hyperparameters
input_dim = 1000  # Vocabulary size
model_dim = 512   # Embedding dimension
num_heads = 8     # Number of attention heads
num_layers = 6    # Number of transformer layers
output_dim = 1000 # Output vocabulary size
seq_length = 10   # Sequence length

# Initialize model, loss function, and optimizer
model = SimpleTransformer(input_dim, model_dim, num_heads, num_layers, output_dim)
criterion_mean = nn.CrossEntropyLoss()
criterion_raw = nn.CrossEntropyLoss(reduction='none')
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Dummy data (batch_size = 2, seq_length = 10)
src = torch.randint(0, input_dim, (seq_length, 2))  # Source sequence
tgt_raw = torch.randint(0, output_dim, (seq_length, 2)) # Target sequence

# Print src and tgt to verify they are the same each time
print("Source sequence (src):")
print("Target sequence (tgt):")

# Forward pass
output = model(src, tgt_raw[:-1, :])  # Exclude the last token in the target sequence

# Reshape output and target for loss calculation
output = output.view(-1, output_dim)
tgt = tgt_raw[1:, :].view(-1)  # Shift target sequence by one

weights = dict(model.named_parameters())
buffers = dict(model.named_buffers())
 
grads_fn = torch.func.grad(compute_loss, has_aux=False, argnums=1)
gs = vmap(grads_fn,
            in_dims=(None, None, None, 1, 1),
            randomness='different')(
                model,
                weights,
                buffers,
                src,
                tgt_raw
            )
```

Everything works without `torch.utils.checkpoint`. Whenever the model contains `torch.utils.checkpoint`, I get the following error after adding `generate_vmap_rule = True` under the `_NoopSaveInputs` class

> RuntimeError: torch.func transforms don't yet support saved tensor hooks. Please open an issue with your use case.
https://github.com/pytorch/pytorch/issues/122729#issuecomment-2681660751

Any plan to make `torch.utils.checkpoint` compatible with `vmap`? With increasing models size, `torch.utils.checkpoint` will be used more often in the foreseeable future, and I think it is necessary to have `torch.utils.checkpoint` compatible with `vmap`.

**Comment by zou3519** [comment]
Feature was implemented. If you have concerns or feedback, please open a new issue.
